source/common/network/listen_socket_impl.h

```diff
@@ -50,5 +50,48 @@ class UdsListenSocket : public ListenSocketImpl {
   UdsListenSocket(const std::string& uds_path);
 };
 
+class AcceptSocketImpl : public AcceptSocket {
+public:
+  AcceptSocketImpl(int fd, Address::InstanceConstSharedPtr&& local_address,
+                   Address::InstanceConstSharedPtr&& remote_address)
+      : fd_(fd), local_address_reset_(false), local_address_(std::move(local_address)),
+        remote_address_(std::move(remote_address)) {}
+  ~AcceptSocketImpl() { close(); }
+
+  // Network::AcceptSocket
+  Address::InstanceConstSharedPtr localAddress() const override { return local_address_; }
+  void resetLocalAddress(Address::InstanceConstSharedPtr& local_address) override {
+    local_address_ = local_address;
+    local_address_reset_ = true;
+  }
+  bool localAddressReset() const override { return local_address_reset_; }
+  Address::InstanceConstSharedPtr remoteAddress() const override { return remote_address_; }
+  void resetRemoteAddress(Address::InstanceConstSharedPtr& remote_address) override {
+    remote_address_ = remote_address;
+  }
+  int fd() const override { return fd_; }
+
+  int takeFd() override {
+    int fd = fd_;
+    fd_ = -1;
+    return fd;
+  }
+
+  void clearReset() override { local_address_reset_ = false; }
```

Comment:
This existence/semantics of the `clearReset()` function is strange to me. It doesn't apply to remote, and it leaves the addresses to what they were reset to. This feels like a hack to me. Can you explain why this is needed and if there is a cleaner way to handle the need?

Reply:
I was never happy with the name of this function either.

This is needed for handing off the accepted socket from one listener to another, which is needed for backwards compatibility in the case where the first listener listens on a redirected port and the second listener matches the original local address but does not bind.

Clearing of the local_address_reset_ bit is needed so that the second listener would not try to repeat the same check again (in an infinite recursion). See the call site in source/server/connection_handler_impl.cc.

This implementation also allows the second listener have and run it's own set of listener filters on the accepted socket it is given.

So the functionality is correct and needed. We should rename this function to lessen the likelihood of confusion in the future, though.

Reply:
OK. I would probably switch reset with "change" and use the following function names (or some variant of them):
1) `changeLocalAddress(...)`
2) `localAddressChanged()`
3) `forgetOriginalLocalAddress()` (This should ASSERT that the local address was changed when called).

Etc.

And more heavily comment the interface. Feel free to iterate on my suggestions above, but that is more clear to me.

Reply:
Just realized that the resetting the local_address_reset_ bit when passing the socket from one listener to another would break the original dst cluster from being able to forward the connection, so I'll have to figure alternative way of not recursing from the second listener. I don't see a need for using an original destination cluster in conjunction of non-bound listeners like this, but I don't like the way this is now anyway, so I'll spend some time on this. I'll try to come up with something that also addresses your concerns here and does not need heavy commenting :-)